### ISSUE TITLE
Make the compiler available online

### ISSUE THREAD

**Issue by krasin** [llvm-dcpu16/llvm-dcpu16#31]
Just a plain pair of input/output windows and "Compile" button would be enough.

**Comment by mgranberry** [comment]
I have a very rough online compiler available here:

http://ferret.granberrys.us:8080/Compiler.html

**Comment by Blei** [comment]
Very cool! The UI is not very polished (and doesn't look correct on firefox apparently, I thought GWT takes care of portability?) but the functionality is great!

**Comment by krasin** [comment]
Great! Do you have the server-side code available?

**Comment by krasin** [comment]
I would recommend to strength your security, because #include directive allows me to read your file system (like ~/.bash_history)

**Comment by krasin** [comment]
There're some suggestions that the compiler lives in chroot. Good, if it's so.

**Comment by mgranberry** [comment]
Yep.  It's offline until I have chroot available on my VPS. :)

**Comment by mgranberry** [comment]
I beefed up security.  It's now running in a near-empty chroot jail with a compiler that can't target X86.  The backend is an ugly mess of Go.   I'll put the code on github once I have a chance to add DoS protection and added some error reporting to the server and added error-handling to the client.

I also have a patched copy of the compiler that disables #include/#import/etc, but I'd like to add support for hosting whole projects, system-installed libraries, and similar features.  Is there a "more right" way to handle this?

**Comment by krasin** [comment]
Yes, it's much better now. I was unable to do anything wrong with it.

I would appreciate if you send us a pull request with your patches. I would put them into `online` branch and will keep that branch up to date.

Open sourcing your Go code is highly appreciated. Nobody cares if it's a mess -- it works and we can do it better, if it's on github.

btw, I like that it's Go. :)

**Comment by mgranberry** [comment]
https://github.com/yrrebnarg/clang/tree/online for the #include-nerfing of clang and
https://github.com/yrrebnarg/dcpu16clangonline for my to-be-renamed web interface.

**Comment by krasin** [comment]
Nice!

Care to make pull request for clang changes? Should be pretty easy.
I will create `online` branch and merge it.

**Comment by krasin** [comment]
Btw, I would not say that your Go code is a mess. It's short and pretty close to idiomatic.

**Comment by krasin** [comment]
Thank you for the pull request!
